File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
private:
    stack<int> stack1;
    stack<int> stack2;

public:
    MinStack() {}

    void push(int val) {
        stack1.push(val);

        if (stack2.empty())
            stack2.push(val);
        else
            stack2.push(min(val, stack2.top()));
    }

    void pop() {
        stack1.pop();
        stack2.pop();
    }

    int top() {
        return stack1.top();
    }

    int getMin() {
        return stack2.top();
    }
};

File: 0155-min-stack/0155-min-stack.py
class MinStack:
    # 3. Min Stack (LeetCode 155)
    def __init__(self):
        # Main stack to store all elements
        self.stack = []
        # Auxiliary stack to store the minimum values at each level
        self.min_stack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        # Push the new minimum onto min_stack (compare with current minimum)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)
        else:
            self.min_stack.append(self.min_stack[-1])

    def pop(self) -> None:
        # Pop from both stacks to keep minimum tracking synchronized
        self.stack.pop()
        self.min_stack.pop()

    def top(self) -> int:
        # Return the top element of the main stack
        return self.stack[-1]

    def getMin(self) -> int:
        # Return the top element of the min stack, which is the current minimum
        return self.min_stack[-1]